Source/Core/VideoBackends/Metal/MTLObjectCache.mm

```diff
@@ -0,0 +1,499 @@
+std::unique_ptr<Metal::ObjectCache> Metal::g_object_cache;
+
+static void SetupDepthStencil(
+    MRCOwned<id<MTLDepthStencilState>> (&dss)[Metal::DepthStencilSelector::N_VALUES]);
+
+Metal::ObjectCache::ObjectCache()
+{
+  m_internal = std::make_unique<Internal>();
+  SetupDepthStencil(m_dss);
+}
+
+Metal::ObjectCache::~ObjectCache()
+{
+}
+
+void Metal::ObjectCache::Initialize(MRCOwned<id<MTLDevice>> device)
+{
+  g_device = std::move(device);
+  g_queue = MRCTransfer([g_device newCommandQueue]);
+  g_object_cache = std::unique_ptr<ObjectCache>(new ObjectCache);
+}
+
+void Metal::ObjectCache::Shutdown()
+{
+  g_object_cache.reset();
+  g_queue = nullptr;
+  g_device = nullptr;
+}
+
+// MARK: Depth Stencil State
+
+// clang-format off
+
+static MTLCompareFunction Convert(CompareMode mode)
+{
+  const bool invert_depth = !g_Config.backend_info.bSupportsReversedDepthRange;
+  switch (mode)
+  {
+  case CompareMode::Never:   return MTLCompareFunctionNever;
+  case CompareMode::Less:    return invert_depth ? MTLCompareFunctionGreater
+                                                 : MTLCompareFunctionLess;
+  case CompareMode::Equal:   return MTLCompareFunctionEqual;
+  case CompareMode::LEqual:  return invert_depth ? MTLCompareFunctionGreaterEqual
+                                                 : MTLCompareFunctionLessEqual;
+  case CompareMode::Greater: return invert_depth ? MTLCompareFunctionLess
+                                                 : MTLCompareFunctionGreater;
+  case CompareMode::NEqual:  return MTLCompareFunctionNotEqual;
+  case CompareMode::GEqual:  return invert_depth ? MTLCompareFunctionLessEqual
+                                                 : MTLCompareFunctionGreaterEqual;
+  case CompareMode::Always:  return MTLCompareFunctionAlways;
+  }
+}
+
+static const char* to_string(MTLCompareFunction compare)
+{
+  switch (compare)
+  {
+  case MTLCompareFunctionNever:        return "Never";
+  case MTLCompareFunctionGreater:      return "Greater";
+  case MTLCompareFunctionEqual:        return "Equal";
+  case MTLCompareFunctionGreaterEqual: return "GEqual";
+  case MTLCompareFunctionLess:         return "Less";
+  case MTLCompareFunctionNotEqual:     return "NEqual";
+  case MTLCompareFunctionLessEqual:    return "LEqual";
+  case MTLCompareFunctionAlways:       return "Always";
+  }
+}
+
+// clang-format on
+
+static void SetupDepthStencil(
+    MRCOwned<id<MTLDepthStencilState>> (&dss)[Metal::DepthStencilSelector::N_VALUES])
+{
+  auto desc = MRCTransfer([MTLDepthStencilDescriptor new]);
+  Metal::DepthStencilSelector sel;
+  for (size_t i = 0; i < std::size(dss); ++i)
+  {
+    sel.value = i;
+    MTLCompareFunction mcompare = Convert(sel.CompareMode());
+    [desc setDepthWriteEnabled:sel.UpdateEnable()];
+    [desc setDepthCompareFunction:mcompare];
+    [desc setLabel:[NSString stringWithFormat:@"DSS %s%s", to_string(mcompare),
+                                              sel.UpdateEnable() ? " ZWE" : ""]];
```

Comment:
What does "ZWE" mean?

Reply:
PCSX2 terminology ("Z Write Enable")
Changed it to `+Write`